app/src/main/java/com/dkanada/gramophone/ui/activities/LoginActivity.java

```diff
@@ -94,41 +111,81 @@ public boolean onOptionsItemSelected(@NonNull MenuItem item) {
+                    ConnectionManager connectionManager = App.getConnectionManager(context, jsonSerializer, logger, httpClient);
+                    connectionManager.Connect(server.getText().toString(), new Response<ConnectionResult>() {
                         @Override
-                        public void onResponse(AuthenticationResult result) {
-                            if (result.getAccessToken() == null) return;
-                            check(context, serverCredentials, result);
+                        public void onResponse(ConnectionResult result) {
+                            App.setApiClient(result.getApiClient());
+                            ServerCredentials serverCredentials = new ServerCredentials();
+                            List<ServerInfo> servers = result.getServers();
+
+                            if (servers.size() < 1) {
+                                Toast.makeText(context, R.string.server_is_unreachable, Toast.LENGTH_SHORT).show();
+                                return;
+                            }
+
+                            serverCredentials.AddOrUpdateServer(servers.get(0));
+                            App.getApiClient().AuthenticateUserAsync(username.getText().toString(), password.getText().toString(), new Response<AuthenticationResult>() {
+                                @Override
+                                public void onResponse(AuthenticationResult result) {
+                                    if (result.getAccessToken() == null) return;
+                                    check(context, serverCredentials, result);
+                                }
+
+                                @Override
+                                public void onError(Exception exception) {
+                                    Log.e(TAG, exception.getMessage());
+                                    Toast.makeText(LoginActivity.this, R.string.authentication_failed, Toast.LENGTH_SHORT).show();
+                                }
+                            });
                         }
                     });
                 }
-            });
+            } else {
+                Intent intent = new Intent(this, SplashActivity.class);
+                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+                startActivity(intent);
+            }
+        }
+    }
+
+    public boolean validate(String mUsername, String mPassword, String mServerAddres) {
+        boolean isValid = true;
+
+        if (TextUtils.isEmpty(mUsername)) {
+            username_layout.setError(getString(R.string.field_cannot_be_empty));
+            isValid = false;
+        } else {
+            username_layout.setError(null);
+        }
+
+        if (TextUtils.isEmpty(mPassword)) {
+            password_layout.setError(getString(R.string.field_cannot_be_empty));
+            isValid = false;
```

Comment:
I think we don't enforce password limits on the server right now, but I think we should leave this anyways because password rules will probably be added soon.